iot-hub/Quickstarts/read-d2c-messages/ReadDeviceToCloudMessages.js

```diff
@@ -1,55 +1,159 @@
+}
+
+/**
+ * Converts an IotHub Connection string into an Event Hubs-compatible connection string.
+ * @param {string} connectionString An IotHub connection string in the format:
+ * `"HostName=<your-iot-hub>.azure-devices.net;SharedAccessKeyName=<KeyName>;SharedAccessKey=<Key>"`
+ * @returns {Promise<string>} An Event Hubs-compatible connection string in the format:
+ * `"Endpoint=sb://<hostname>;EntityPath=<your-iot-hub>;SharedAccessKeyName=<KeyName>;SharedAccessKey=<Key>"`
+ */
+async function convertIotHubToEventHubsConnectionString(connectionString) {
+  const { HostName, SharedAccessKeyName, SharedAccessKey } = parseConnectionString(
+    connectionString
+  );
+
+  // Verify that the required info is in the connection string.
+  if (!HostName || !SharedAccessKey || !SharedAccessKeyName) {
+    throw new Error(`Invalid IotHub connection string.`);
+  }
+
+  //Extract the IotHub name from the hostname.
+  const [iotHubName] = HostName.split(".");
+
+  if (!iotHubName) {
+    throw new Error(`Unable to extract the IotHub name from the connection string.`);
+  }
+
+  // Generate a token to authenticate to the service.
+  const token = generateSasToken(
+    `${HostName}/messages/events`,
+    SharedAccessKey,
+    SharedAccessKeyName,
+    5 // token expires in 5 minutes
+  );
+  const connectionOptions = {
+    transport: "tls",
+    host: HostName,
+    hostname: HostName,
+    username: `${SharedAccessKeyName}@sas.root.${iotHubName}`,
+    port: 5671,
+    reconnect: false,
+    password: token,
+  };
+
+  const connection = new Connection(connectionOptions);
+  await connection.open();
+
+  // Create the receiver that will trigger a redirect error.
+  const receiver = await connection.createReceiver({
+    source: { address: `amqps://${HostName}/messages/events/$management` },
+  });
+
+  return new Promise((resolve, reject) => {
+    receiver.on(ReceiverEvents.receiverError, (context) => {
+      const error = context.receiver && context.receiver.error;
+      if (isAmqpError(error) && error.condition === "amqp:link:redirect") {
+        const hostname = error.info && error.info.hostname;
+        if (!hostname) {
+          reject(error);
+        } else {
+          resolve(
+            `Endpoint=sb://${hostname}/;EntityPath=${iotHubName};SharedAccessKeyName=${SharedAccessKeyName};SharedAccessKey=${SharedAccessKey}`
+          );
+        }
+      } else {
+        reject(error);
+      }
+      connection.close().catch(() => {
+        /* ignore error */
+      });
+    });
+  });
+}
 
 var printError = function (err) {
   console.log(err.message);
 };
 
 // Display the message content - telemetry and properties.
 // - Telemetry is sent in the message body
-// - The device can add arbitrary application properties to the message
+// - The device can add arbitrary properties to the message
 // - IoT Hub adds system properties, such as Device Id, to the message.
 var printMessage = function (message) {
```

Comment:
Since we're passing this directly as `processEvents`, `message` is an array of messages.
```suggestion
var printMessages = function (messages) {
  for (const message of messages) {
  /* Log message details */
  }
 };
```

Reply:
Good catch, fixed with https://github.com/Azure-Samples/azure-iot-samples-node/pull/38/commits/c762748e69b3fdb3bc5fb931fe9d5d708eb27a3c